### ISSUE TITLE
Consider a Media Type for CSAF

### ISSUE THREAD

**Issue by sonnyvanlingen** [oasis-tcs/csaf#639]
As discussed during the OASIS CSAF TC Monthly Meeting on 2023-07-26:

The CSAF 2.0 specification does not provide a predefined Media sub type or MIME type for CSAF documents. This means that consumers of CSAF have to rely on context or to parse the document to understand its nature.

It is suggested to define a MIME type for CSAF documents and register it at IANA. This proposed change can help in reducing processing time and system load by CSAF consumers. It can also help with the exchange of CSAF documents across systems or community boundaries, without requiring standardization of all distribution methods in the CSAF standard. 

I look forward to hear the TC's insights and feedback.

BRs,

Sonny van Lingen
Huawei PSIRT

**Comment by sthagen** [comment]
I am in favor of this proposal and amend by suggesting `application/csaf+json` as the mime type to expect for CSAF advisory files. I also added the CSAF 2.1 label as I see the same need when we introduce an API, which we intend to do (in my understanding) for CSAF version 2.1. We can always remove a label, but I think this is the right signal.

Maybe we can best start thinking and discussing about this topic per the following table:

| MIME type                                  | Extension                                  | Description                                 |
|:-------------------------------------------|:-------------------------------------------|:--------------------------------------------|
| application/csaf+json<br>&nbsp;            | .csaf,<br>.csaf.json                       | CSAF files<br>&nbsp;                        |
| application/csaf-aggregator+json<br>&nbsp; | .csaf-aggregator,<br>.csaf-aggregator.json | CSAF aggregator information files<br>&nbsp; |
| application/csaf-provider+json<br> &nbsp;  | .csaf-provider,<br>.csaf-provider.json     | CSAF provider information files <br>&nbsp;  |

**Comment by sthagen** [comment]
During the 2023-09-27 meeting of the TC the members approved the motion to add the media types as per Stefan's proposal <https://github.com/oasis-tcs/csaf/issues/639#issuecomment-1666415334>

**Comment by mipfu** [comment]
I have a clear position here
.CSAF is a content type
.json is a format-standard type

Media types should be primarly used to identify content type through the last DOT-Prefix.

Using a format-standard type at the end violates the purpose of the CSAF standard by wanting to support which front-end technology (Viewers). Let the User decide which Viewer he wants to use for CSAF! Suffixing with .json eliminates this freedom to some degree and is out of scope of CSAF Standard.

**Comment by bernhardreiter** [comment]
CSAF 2.0 uses `.json` file suffixes.
**A change to other file suffixes is not advisable,**
as there are drawbacks and no practical advantages.
The registration of media-types does not do much harm,
but also does not promise significant benefits.

## Details
We were asked by @tschmidtb51 to check into practical consequences of registering Media-Types and two variant of file suffixes from an implementer's perspective. Within a 10 hours limited research we have found that common operating systems (GNU/Linux, MacOS, Windows) and web servers (Apache2, freenginx) will not easily accept new suffixes as default and do not handle double-dot-suffixes easily.

A comparison with the SARIF standard shows that `.sarif` is not accepted as default in nginx for many years  and not even if requested explicitly within a month. The two file suffix variants there are meant to support editing applications for humans. The situation is  (from our perspective) different for CSAF documents:

 * With specific APIs, e.g. for a CSAF management system,  the media types do not really matter, as result types per end-point will be known to the calling application.
 *  For human applications, CSAF 2.1 will offer a `$schema` property that will allow editors' JSON functionality to know more about the file.  So a detection per file suffixes is unnecessary for this use case, too.

The drawbacks of moving away from the CSAF 2.0 file suffixes include changing existing implementations, the work for media-type registration and maintenance. And work trying to get media-types and file names into defaults for operating systems and web servers.

Two experienced software engineers also did not come up with significant advantages. A potential pro argument when stopping bots via content delivery networks was not further explored.


### Comparison with SARIF

* Trying to get `.sarif` into nginx, Pull-Request: <https://github.com/nginx/nginx/pull/592>,
  opened 2025-03-24. Not done a months later. 
  -> it is unlikely to get a new suffix into nginx defaults.

* Asked the SARIF TC about the rationale for double dot file name extensions in <https://github.com/oasis-tcs/sarif-spec/issues/683>. Result: Implementations have to deal with both endings and there is a choice.
  -> Suboptimal for good standard practice, because of unnecessary variations.

**Comment by elear** [comment]
What you are discussing is a UX issue.

Having a suffix is good if you are going to have web browser interactions with file system in Javascript (for instance), where suffixes identify valid files for selection.  Because JSON is used for EVERYTHING, a suffix might be useful to help the user avoid uploading inappropriate files.  However, if a JSON file **can** be valid, then users may be prohibited by the UX from otherwise selecting valid files.

Choose wisely.

**Comment by bernhardreiter** [comment]
> if you are going to have web browser interactions with file system in Javascript (for instance), where suffixes identify valid files for selection. [..] a suffix might be useful to help the user avoid uploading inappropriate files.

Yes, this is an argument for changing towards using unique single suffixes for each CSAF file. A web application (or page) that prompts the user to select a file for uploading can use the [accept property of input](https://html.spec.whatwg.org/multipage/input.html#attr-input-accept) or the [equivalent API](https://developer.mozilla.org/de/docs/Web/API/HTMLInputElement/accept) to give users _a hint of what file types will be accepted_.

But subsequently, the web application has to check the file type by itself anyway. From the  former linked `accept property documentation` :

_Authors are reminded that, as usual, data received from a client should be treated with caution, as it may not be in an expected format even if the user is not hostile and the user agent fully obeyed the accept attribute's requirements._

Therefore: the first thing a web application for CSAF 2.1 files will do is to check the `$schema` property and can then inform the users if this is not the expected CSAF file. Advanced text editors do the same. All CSAF files already have significant names, so that a human has a high probability to recognize the one they want to select for uploading/opening. A move to new suffixes seem to only provide a quite small advantage here.

In case of desktop operating systems, it is more likely that a connection of opening `.json` files with an advanced code editors is already in place, while a connection to a new suffix is hard to get installed. So many user may profit from getting the CSAF file directly in their editor by default, if the suffix situation stays as it is. Those little UX differences have to be summed up and weighted against each other.  The large drawbacks for new suffixes is that existing implementations (and humans mental recognition) have to be changed. And there is significant work to try to get Media Types and suffixes into operating systems', web servers' and other applications' defaults - which has a low chance of succeeding. Thus we can to the conclusion that in total from an implementor's view including the UX aspects, it is better to keep the suffix situation with `.json` as it is.

**Comment by elear** [comment]
Good answer.  I would go a little bit rather and actually perform a schema validation.
